src/thunderbots/software/ai/navigator/obstacle.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include "geom/util.h"
+#include "ai/world/world.h"
+
+constexpr double DEFAULT_AVOID_DIST = 0.15;  // meters
+
+class RobotObstacle
+{
+   public:
+    /**
+     * Constructor.
+     *
+     * @param r Robot to create obstacle from.
+     */
+    RobotObstacle(Robot& r, double avoid_dist);
+
+    /**
+     * Returns the violation distance of the given point inside the obstacle
+     * boundary; 0.0 if no violation.
+     *
+     * @param p point to check if violating
+     *
+     * @return Violation distance, defined as distance from the nearest boundary
+     *         to the given point; 0.0 if no violation occured.
+     */
+    double getViolation(Point& p);
+
+    /**
+     * Returns the closest Point on the obstacle boundary to the given Point that
+     * is not inside the boundary. If the given Point does not violate the boundary,
+     * the given Point will be returned.
+     *
+     * @param p Point to check for
+     * @return A point on the boundary nearest to p if p violates the boundary, otherwise
+     *         returns p
+     */
+    Point getNearestValidPoint(Point& p);
+
+    /**
+     * TODO: This function may have to be rewritten depending on our needs later.
+     * Checks if the given Robot will collide with the obstacle based on the current
+     * velocities.
+     *
+     * @param r Robot to check against.
+     * @return true if collision possible, false otherwise.
+     */
+    bool willCollide(Robot& r);
+
+   private:
+    /**
+     * Represents the physical footprint of the robot, with
+     * an additional avoid distance.
+     */
+    Circle boundary;
+
+    /**
+     * Segment extending from robot in direction of robot travel with length
+     * proportional to current velocity. Used for collision avoidance.
+     */
+    Seg velocity;
```

Comment:
I don't remember seeing the `Seg` class go through code review (my apologies if it did and I missed it), but could we rename it to `Segment`? It's four extra characters and makes things much more clear (ex. is it a Segfault??)

Reply:
The Seg and Rect stuff was ported from the old AI, in @jordanzliu's review (#29). Should probably be done as a separate review, though; this review already changes a bunch of files.

Reply:
Fair enough, opened https://github.com/UBC-Thunderbots/Software/issues/40 to track this issue.

Reply:
wontfix